### PR TITLE
CUDA: backwards pass for misc. ops, add tests

### PR DESCRIPTION
**Pull request by JohannesGaessler**
List of changes:

* Extended the interface of the `SOFT_MAX` backward pass to be equivalent to the forward pass. Support for `scale` is implemented, support for `max_bias` is not implemented. There is no need for `mask` since it does not affect the gradients of the forward pass input beyond its effects on the forward pass output. Added CUDA support for the backward pass.
* Fixed the CUDA implementation of `CROSS_ENTROPY_LOSS` not working correctly for large vocabulary sizes.
* Implemented broadcasting of dimensions 2 and 3 for `OUT_PROD` to support the backward pass for GQA.
* Added CUDA support for the `GET_ROWS` backward pass.
* Added CUDA support for the `SILU` backward pass.
* Added CUDA support for the `RMS_NORM` backward pass.
* Added tests to `test-backend-ops` to assert the consistency of the CPU and CUDA implementations.
* Refactored the dedicated functions for backward pass construction to always have the gradients as the first argument.

### CODE REVIEW

File: ggml/src/ggml-cuda/unary.cu [hunk 1]
```diff
@@ -20,7 +20,7 @@ static __global__ void step_f32(const float * x, float * dst, const int k) {
     dst[i] = x[i] > 0.0f;
 }
 
-static __global__ void gelu_f32(const float * x, float * dst, const int k) {
+static __global__ void gelu_f32(const float * __restrict__ x, float * __restrict__ dst, const int k) {
```
Comment (slaren) [inline ggml/src/ggml-cuda/unary.cu:23]: I don't think not sure that restrict is correct in these functions, since they can run in-place, which would make both pointers point to the same address.

File: ggml/src/ggml-cpu/ggml-cpu.cpp [hunk 2]
```diff
@@ -403,6 +403,16 @@ static bool ggml_backend_cpu_device_supports_op(ggml_backend_dev_t dev, const st
                 op->type != GGML_TYPE_IQ1_M; // missing type_traits.from_float
         case GGML_OP_MUL_MAT:
             return src1->type == GGML_TYPE_F32 || src1->type == ggml_get_type_traits_cpu(src0->type)->vec_dot_type;
+        case GGML_OP_SOFT_MAX_BACK: {
+            if (op->src[0]->type != GGML_TYPE_F32 || op->src[1]->type != GGML_TYPE_F32) {
+                return false;
+            }
+            float max_bias = 0.0f;
+
+            memcpy(&max_bias, (const float *) op->op_params + 1, sizeof(float));
```
Comment (ggerganov) [inline ggml/src/ggml-cpu/ggml-cpu.cpp:412]: At some point we should make the op params get/set more strongly typed (e.g. via get/set functions for each one). Just an idea, no action required for this PR.

File: ggml/src/ggml-cuda/unary.cu (hunk 1 above)
Comment (JohannesGaessler) [reply ggml/src/ggml-cuda/unary.cu:23]: ggml ops can only be safely run in-place if both the source and destination pointers have the same memory layout. Otherwise you would run into issues with race conditions. The use of restrict allows the compiler to reorder memory accesses during optimization under the assumption that writes to one pointer do not affect reads from another pointer. But the compiler can never optimize the code to write the result before it reads the inputs. So the use of restrict is safe with in-place ops.

My opinion is that the use of restrict should be the default. I don't feel strongly about this though and would also be fine with making the use of restrict the exception for cases where it can be demonstrated to be safe and useful.

File: ggml/src/ggml-cpu/ggml-cpu.cpp (hunk 2 above)
Comment (JohannesGaessler) [reply ggml/src/ggml-cpu/ggml-cpu.cpp:412]: I think the correct way to do this would be to define per-op structs for the op params.

File: ggml/src/ggml-cuda/unary.cu (hunk 1 above)
Comment (JohannesGaessler) [reply ggml/src/ggml-cuda/unary.cu]: I disagree but I removed the use of restrict on all ops that can be used in-place, either explicitly through the `ggml.h` API or implicitly through `ggml_op_can_inplace`. On a somewhat related note, are there plans to eventually remove the `_inplace` functions in `ggml.h` and to have the logic regarding in-place operations be handled completely automatically?

Comment (slaren) [reply ggml/src/ggml-cuda/unary.cu]: In ggml-alloc in-place is used to save memory, but in-place operations can also be useful to modify a static tensor. For example, in llama.cpp an inplace ROPE is used to perform the contexts shifts. Without inplace, this would require a normal rope followed by a cpy. In principle it would be possible to optimize away unnecessary cpy operations by making the previous operation inplace automatically, but that's not implemented, and probably not worth the complexity. So at the moment in-place operations still have a use.

**Review by slaren — APPROVED** [review]

File: ggml/src/ggml-cuda/cross-entropy-loss.cu [hunk 3]
```diff
@@ -119,48 +113,77 @@ void ggml_cuda_cross_entropy_loss(ggml_backend_cuda_context & ctx, ggml_tensor *
...
 
-    const dim3 blocks_dim(CUDA_CROSS_ENTROPY_LOSS_BLOCK_SIZE, 1, 1);
-    const dim3 blocks_num((nrows + CUDA_CROSS_ENTROPY_LOSS_BLOCK_SIZE - 1) / CUDA_CROSS_ENTROPY_LOSS_BLOCK_SIZE, 1, 1);
-    const int shmem = 2*CUDA_CROSS_ENTROPY_LOSS_BLOCK_SIZE*ne00*sizeof(float);
+    const dim3 blocks_dim(WARP_SIZE, 1, 1);
+    const dim3 blocks_num(nrows, 1, 1);
+    const int nbytes_shared = ne00*sizeof(float);
+
+    const int id    = ggml_cuda_get_device();
+    const int smpbo = ggml_cuda_info().devices[id].smpbo;
```
Comment (ggerganov) [inline ggml/src/ggml-cuda/cross-entropy-loss.cu:121]:
```suggestion
    const size_t nbytes_shared = ne00*sizeof(float);

    const int    id    = ggml_cuda_get_device();
    const size_t smpbo = ggml_cuda_info().devices[id].smpbo;
```

File: ggml/src/ggml-cuda/cross-entropy-loss.cu [hunk 4]
```diff
@@ -119,48 +113,77 @@ void ggml_cuda_cross_entropy_loss(ggml_backend_cuda_context & ctx, ggml_tensor *
...
 
     const dim3 blocks_dim(WARP_SIZE, 1, 1);
     const dim3 blocks_num(nrows, 1, 1);
-    const int shmem = ne00*sizeof(float);
-
-    cross_entropy_loss_back_f32<<<blocks_num, blocks_dim, shmem, stream>>>(src0_d, src1_d, opt0_d, dst_d, ne00);
+    const int nbytes_shared = ne00*sizeof(float);
+
+    const int id    = ggml_cuda_get_device();
+    const int smpbo = ggml_cuda_info().devices[id].smpbo;
```
Comment (ggerganov) [inline ggml/src/ggml-cuda/cross-entropy-loss.cu:175]:
```suggestion
    const size_t nbytes_shared = ne00*sizeof(float);

    const int    id    = ggml_cuda_get_device();
    const size_t smpbo = ggml_cuda_info().devices[id].smpbo;
```

File: ggml/src/ggml-cuda/norm.cu [hunk 5]
```diff
@@ -127,12 +130,63 @@ static __global__ void rms_norm_f32(const float * x, float * dst, const int ncol
...
+
+template <int block_size>
+static __global__ void rms_norm_back_f32(
+        const float * grad, const float * xf, float * dst, const int ncols, const float eps) {
+    const int row = blockIdx.x*blockDim.y + threadIdx.y;
+    const int tid = threadIdx.x;
+
+    grad += row*ncols;
+    xf   += row*ncols;
+    dst  += row*ncols;
```
Comment (ggerganov) [inline ggml/src/ggml-cuda/norm.cu:145]: Is there risk of integer overflow here? Maybe to be safe:

```suggestion
    grad += int64_t(row)*ncols;
    xf   += int64_t(row)*ncols;
    dst  += int64_t(row)*ncols;
```

Same comment in a few other places where we multiply `row*ncols`.

**Review by ggerganov — APPROVED** [review]
